### ISSUE TITLE
podman commit --include-volumes does only work with value true documentation needs to reflect that 

### ISSUE THREAD

**Issue by 6ramr** [containers/podman#15387]
/kind bug

**Description**
documentation at https://docs.podman.io/en/latest/markdown/podman-commit.1.html reads:

--include-volumes

Include in the committed image any volumes added to the container by the --volume or --mount OPTIONS to the [podman create](https://docs.podman.io/en/latest/markdown/podman-create.1.html) and [podman run](https://docs.podman.io/en/latest/markdown/podman-run.1.html) commands.
The default is false.

**Steps to reproduce the issue:**

1. podman commit --include-volumes [options] container [image]

2. podman run --name [name] [options] [image]

3. podman exec -it [options] [container] /bin/ls _path-to-volume_

4. is not altered

**Describe the results you received:**

using the option does not include the altered volumes

**Describe the results you expected:**

does include altered volumes, or throws error because of value not specified

__OR__ at least:

documentation should read

--include-volumes=_bool_ (true/false)

Include in the committed image any volumes added to the container by the --volume or --mount OPTIONS to the [podman create](https://docs.podman.io/en/latest/markdown/podman-create.1.html) and [podman run](https://docs.podman.io/en/latest/markdown/podman-run.1.html) commands.
The default is false.

__like__ documentation at https://docs.podman.io/en/latest/markdown/podman-run.1.html

--cgroupns=mode

Set the cgroup namespace mode for the container.

host: use the host’s cgroup namespace inside the container.
container:id: join the namespace of the specified container.
private: create a new cgroup namespace.
ns:path: join the namespace at the specified path.
If the host uses cgroups v1, the default is set to host. On cgroups v2, the default is private.

**Comment by 6ramr** [comment]
don't know how to add labels..
should be bug and documentation.

**Comment by mheon** [comment]
So, from your description, it sounds like `--include-volumes=true` works? This sounds like a bug...

**Comment by 6ramr** [comment]
Tbh, i was very stressed out today at work. And I cannot confirm that any configuration works. I think I wrote that report prematurely.

So, no matter whether I specify --include-volumes or --include-volumes=true, the image I committed didn't have the files when I ran it on another machine. And there was no error message.

Thank you for having a look at this.

I should have gone deeper but I ran out of time and the earliest I will have another look at it will be 29th of August, since I'm on a direly needed vacation.

Mit freundlichen Grüßen,
Julian Steiner

________________________________
Von: Matt Heon ***@***.***>
Gesendet: Freitag, 19. August 2022, 17:19
An: containers/podman ***@***.***>
Cc: 6ramr ***@***.***>; Author ***@***.***>
Betreff: Re: [containers/podman] podman commit --include-volumes does only work with value true documentation needs to reflect that  (Issue #15387)


So, from your description, it sounds like --include-volumes=true works? This sounds like a bug...

**Comment by mheon** [comment]
OK, so `--include-volumes` does not work at all. At least our CLI is consistent.

Will tag as a bug.

**Comment by flouthoc** [comment]
Oh so I am not sure about this flag, `docker commit` does not supports this and why should image direct about how the container has to be started i.e which mounts should be included ? The implementation of flag sets the `volume` field of image-spec and does nothing else, I image should not secretly tell container managers to mount a volume.

If user wants content of volume in the image I think user should copy content from host.

Maybe we should drop this option. @mheon @containers/podman-maintainers WDYT ?

**Comment by 6ramr** [comment]
I'm doing some modifications on a pod with multiple containers and shared volumes. This should push the container data to the image.

Containers mount predefined volumes from the images anyway. There is even an option to suppress that behaviour.

Granted, stopping, unmounting and copying is is indeed an option, but simply specifying to include the volume-data is simply more elegant and user-friendly.

Mit freundlichen Grüßen,
Julian Steiner
________________________________

**Comment by flouthoc** [comment]
@6ramr The problem is that the newly built image will only work as expected with podman and not with other container managers which don't implement mounting volumes directed from the image-spec of a container image, but again this just my opinion and other might disagree so maybe lets wait for others to take a look.

**Comment by 6ramr** [comment]
Yeah. I guess we're the only crazies who are actually actively developing a bash based podman container management tool .. but it's podman under the hood .

Mit freundlichen Grüßen,
Julian Steiner
________________________________

**Comment by rhatdan** [comment]
Who added --include-volumes flag?
buildah commit does no have it.

**Comment by rhatdan** [comment]
```
    Add --include-volumes flag to 'podman commit'
    
    The 'docker commit' will never include a container's volumes when
    committing, without an explicit request through '--change'.
    
    Podman, however, defaulted to including user volumes as image
    volumes.
    
    Make this behavior depend on a new flag, '--include-volumes',
    and make the default behavior match Docker.
    
    Signed-off-by: Matthew Heon <matthew.heon@pm.me>
```

**Comment by rhatdan** [comment]
@mheon looks like it was added in 2019.

**Comment by mheon** [comment]
@flouthoc The intent is to squash the image's contents *into the image*, not to bake the volumes themselves into the image. I don't see how that can break other container managers. The original implementation probably got lost at some point along the way - 2019 means it might actually predate `podman volume` support, for example.

@rhatdan This was an old feature request I vaguely recall. Intent was to make shipping the whole contents of a container possible (way back before we had `volume export` and `volume import`). After committing with `--include-volumes`, you can create a new container with the committed image and mount volumes at the same places, and copy-up will effectively restore your volumes by copying appropriate data from the image into them. Makes a lot of sense to me.

**Comment by mheon** [comment]
Hm. Or I might be thinking about an entirely separate feature request from back then...

Looking at it more, this could be just a legacy-compat feature from us discovering that we implemented commit wrong.

I still don't share the worries that @flouthoc has; mounting volumes is part of the image spec for a reason, and I don't see how using it breaks other container managers. I see no reason not to fix this.

But I have that other old feature request stuck in my head now, and it seems like a really good idea... Maybe I should throw it in as an RFE

**Comment by flouthoc** [comment]
@mheon I think it allows external image creators to access host data. Consider a container image
* Which secretly mounts `/etc/someimportantdir` into container.
* Entrypoint binary reads data and post it to external dest.

In above scenario the user who is trying to use the image is unware of the image config which is secretly mounting the host data and posting it to external dest. My concern is that here `image` gets to decide if it wants to read data from host or not.

**Comment by flouthoc** [comment]
I even think that this should be removed from upstream image-spec but i could be missing things so this just opinion.

**Comment by mheon** [comment]
@flouthoc We are not encoding the destination of the volume, just that a volume was mounted as a specific location in the image.

**Comment by mheon** [comment]
This also does not apply to bind mounts, to be clear. This is Podman-managed volumes only, AFAIK.

**Comment by nalind** [comment]
So if `podman create`'s (or `podman run`'s) `-v` option was used with a named source volume (but not a directory path on the host), `podman commit --include-volumes` just adds the equivalent of a `-c "VOLUME ..."` for each location in the container that was used as the mountpoint for a volume?

**Comment by mheon** [comment]
That appears to have been the original intent of the PR. I think `podman commit` originally did this automatically, which was not what Docker did, so when we reverted to the Docker behavior we added this flag to retain the original Podman functionality.

**Comment by mheon** [comment]
How can it access host data? I fail to understand.

**Comment by flouthoc** [comment]
> How can it access host data? I fail to understand.

Not as of now but i think if we complete `--include-volumes` to as it is written in docs and as it is requested in this github issue then it could right ?

**Comment by mheon** [comment]
If the user requests that volume content be cached into the image (again, to be clear, Libpod-managed volumes, not bind mounts), and the volumes include confidential data, I don't see how that's any more of a problem than what we do right now. I could do a `podman volume export` on the volume and get the same effect.

**Comment by rhatdan** [comment]
@flouthoc @mheon What is the state of this issue?

**Comment by mheon** [comment]
Seems like it's more of an RFE than a bug at this point, though the flag admittedly does not function at present.

**Comment by rhatdan** [comment]
@flouthoc any movement on this one?
